packages/apollo-engine-reporting/src/agent.ts

```diff
@@ -349,30 +307,37 @@ export class EngineReportingAgent<TContext = any> {
     });
 
     const statsReportKey = `# ${operationName || '-'}\n${signature}`;
-    if (!this.report.tracesPerQuery.hasOwnProperty(statsReportKey)) {
-      this.report.tracesPerQuery[statsReportKey] = new Traces();
-      (this.report.tracesPerQuery[statsReportKey] as any).encodedTraces = [];
+    if (!report.tracesPerQuery.hasOwnProperty(statsReportKey)) {
+      report.tracesPerQuery[statsReportKey] = new Traces();
+      (report.tracesPerQuery[statsReportKey] as any).encodedTraces = [];
     }
     // See comment on our override of Traces.encode inside of
     // apollo-engine-reporting-protobuf to learn more about this strategy.
-    (this.report.tracesPerQuery[statsReportKey] as any).encodedTraces.push(
+    (report.tracesPerQuery[statsReportKey] as any).encodedTraces.push(
       encodedTrace,
     );
-    this.reportSize += encodedTrace.length + Buffer.byteLength(statsReportKey);
+    this.reportSizes[schemaHash] +=
+      encodedTrace.length + Buffer.byteLength(statsReportKey);
 
     // If the buffer gets big (according to our estimate), send.
     if (
       this.sendReportsImmediately ||
-      this.reportSize >=
+      this.reportSizes[schemaHash] >=
         (this.options.maxUncompressedReportSize || 4 * 1024 * 1024)
     ) {
-      await this.sendReportAndReportErrors();
+      await this.sendReportAndReportErrors(schemaHash);
     }
   }
 
-  public async sendReport(): Promise<void> {
-    const report = this.report;
-    this.resetReport();
+  public async sendAllReports(): Promise<void> {
+    await Promise.all(
```

Comment:
My first reaction is that Promise.all is problematic here because it doesn't show *all* errors, just the first one. I think you figured that out though, which is why this method is never called. So please remove it if I'm understanding correctly.

Reply:
it is called from `ApolloServer#stop()`

Reply:
Hmm, it does feel like that should block until all sends are done or have failed, rather than potentially finishing while there are still some in-flight sends.